push/src/gateway.cc

```diff
@@ -15,6 +16,31 @@ namespace prometheus {
 static const char CONTENT_TYPE[] =
     "Content-Type: text/plain; version=0.0.4; charset=utf-8";
 
+class CurlWrapper {
+public:
+  CurlWrapper() {
+    curl_ = nullptr;
+  }
+  ~CurlWrapper() {
+    std::lock_guard<std::mutex> l(mutex_);
+    if (curl_) {
+      curl_easy_cleanup(curl_);
+    }
+  }
+
+  CURL *curl() {
+    std::lock_guard<std::mutex> l(mutex_);
+    if (!curl_) {
+      curl_ = curl_easy_init();
+    }
+    return curl_;
+  }
+
+private:
+  CURL *curl_;
+  std::mutex mutex_;
```

Comment:
The mutex here is misplaced and should be moved into `Gateway` to protect from concurrent access from `Gateway::Push()`